### PR TITLE
add `element.evaluate`

### PR DESCRIPTION
**Pull request by teawinzero**

### CODE REVIEW

File: src/element_handle.ts [hunk 1]
```diff
@@ -369,4 +383,78 @@ export class ElementHandle {
...
+   * )
+   * ```
+   */
+  async evaluate<E extends Element, R extends AnyArray, T = unknown>(
+    func: ElementEvaluateFunction<E, R, T> | string,
+    evaluateOptions?: ElementEvaluateOptions<R>,
+  ): Promise<T> {
+    const { object } = await this.#celestial.DOM.resolveNode({
+      nodeId: this.#id,
+    });
```
Comment (lino-levan) [inline src/element_handle.ts:411]: This needs a retry around it

File: src/element_handle.ts [hunk 2]
```diff
@@ -369,4 +383,78 @@ export class ElementHandle {
...
+    }
+
+    const { result, exceptionDetails } = await this.#celestial.Runtime
+      .callFunctionOn({
+        functionDeclaration: func.toString(),
+        objectId: object.objectId,
+        arguments: args,
+        awaitPromise: true,
+        returnByValue: true,
+      });
```
Comment (lino-levan) [inline src/element_handle.ts:438]: This needs a retry

Comment (teawinzero) [reply src/element_handle.ts:438]: If wrapped in this way, a DeadlineError will appear after ~55 seconds of execution
```ts
const { result, exceptionDetails } = await retryDeadline(
  this.#celestial.Runtime.callFunctionOn({ ... }),
  this.#page.timeout, // 10_000
);
```
```ts
await element.evaluate(async () => {
  await new Promise((resolve) => setTimeout(resolve, 120_000));
})
```
RetryError: Retrying exceeded the maxAttempts (5).
Caused by DeadlineError: Deadline

maybe add as a `timeout` option and only use `deadline`?
```ts
interface ElementEvaluateOptions<T> {
  args: Readonly<T>;
  timeout?: number;
}

const evaluatePromise = this.#celestial.Runtime
  .callFunctionOn({
    functionDeclaration: func.toString(),
    objectId: object.objectId,
    arguments: args,
    awaitPromise: true,
    returnByValue: true,
  });

const { result, exceptionDetails } = await deadline(
  evaluatePromise,
  evaluateOptions?.timeout ?? this.#page.timeout,
);
```

Comment (lino-levan) [reply src/element_handle.ts:438]: I think it's unrealistic that user code being evaluated takes more than 10 seconds to run. If it is, they can just change the timeout. I don't see this being an issue.

The real issue is actually the CDP protocol randomly never responds. This causes hangs that are impossible to reproduce. All CDP protocol runs should be retried and deadlined.

File: src/element_handle.ts [hunk 3]
```diff
@@ -1,6 +1,8 @@
+/// <reference lib="dom" />
+
 import { deadline } from "https://deno.land/std@0.205.0/async/deadline.ts";
 
-import { Celestial } from "../bindings/celestial.ts";
+import { Celestial, Runtime_CallArgument } from "../bindings/celestial.ts";
```
Comment (lino-levan) [inline src/element_handle.ts:5]:
```suggestion
import { Celestial, type Runtime_CallArgument } from "../bindings/celestial.ts";
```

File: src/element_handle.ts [hunk 4]
```diff
@@ -1,6 +1,8 @@
+/// <reference lib="dom" />
```
Comment (lino-levan) [inline src/element_handle.ts:1]: This is a little scary. Do we need this?

Comment (teawinzero) [reply src/element_handle.ts:1]: I checked, we can delete it

**Comment by lino-levan** [comment]
@lowlighter we're getting pretty consistent issues with booting the binary recently on windows. Any ideas on why that might be happening?

**Comment by teawinzero** [comment]
I can assume the problem is in the shared folder when starting the process

This can be solved using the `--user-data-dir=` argument.
Note: Files with a total size of ~60MB are created in the specified folder. Which will need to be cleaned later if this is not a permanent profile.

https://github.com/microsoft/playwright/blob/11bf96fe98decb99a6abd08578b4d27c06e49cd9/packages/playwright-core/src/server/chromium/chromium.ts#L282
```ts
chromeArguments.push(`--user-data-dir=${userDataDir}`);
```

https://github.com/microsoft/playwright/blob/11bf96fe98decb99a6abd08578b4d27c06e49cd9/packages/playwright-core/src/server/browserType.ts#L161
```ts
userDataDir = await fs.promises.mkdtemp(path.join(os.tmpdir(), `playwright_${this._name}dev_profile-`));
```

**Comment by lino-levan** [comment]
Wow! This has been a consistent issue we've been facing forever. This definitely looks to be it. Would you be willing to PR this or should I?

**Comment by lino-levan** [comment]
Should just be a `Deno.makeTempDir` + passing this argument.

**Comment by lino-levan** [comment]
@teawinzero could you merge this into main?

File: src/element_handle.ts (hunk 4 above)
Comment (lino-levan) [reply src/element_handle.ts:1]: Does this implicitly add the dom reference to downstream consumers? If so, we can't really do this.

Comment (teawinzero) [reply src/element_handle.ts:1]: I have no idea how to solve this

Comment (lino-levan) [reply src/element_handle.ts:1]: We don't need to strongly type the argument of the evaluate function. Making it `(element: unknown, ...` would be preferable to forcing all downstream consumers to use the dom reference.

**Review by lino-levan — APPROVED** [review]
LGTM. Thanks!

**Pull request merged**